### PR TITLE
Accept input tensor with 0-dim batch size for MultiLabelMarginLoss

### PR DESCRIPTION
**Pull request by v0dro**
Fix for one of the layers listed in https://github.com/pytorch/pytorch/issues/12013 or https://github.com/pytorch/pytorch/issues/38115

### CODE REVIEW

**Comment by v0dro** [comment]
@ngimel request review.

**Comment by v0dro** [comment]
@ngimel could you please have a look?

File: test/test_nn.py [hunk 1]
```diff
@@ -10351,6 +10351,27 @@ def test_ReflectionPad_empty(self, device):
...
             mod(inp)
 
+    @onlyOnCPUAndCUDA
+    def test_MultiLabelMarginLoss_empty(self, device):
+        mod = torch.nn.MultiLabelMarginLoss()
+        x = torch.randn(0, 10, requires_grad=True)
+        y = torch.ones(0, 10).type(torch.long)
+
+        out = mod(x, y)
+        out.sum().backward()
```
Comment (mruberry) [inline test/test_nn.py:10361]: Check the value of out and x.grad, too.

**Comment by mruberry** [comment]
Hey v0dro!

I made a comment about the test coverage. What about CUDA, by the way? Does it already handle inputs with a zero size batch dim?

**Comment by v0dro** [comment]
@mruberry could you please review this again?

File: test/test_nn.py [hunk 2]
```diff
@@ -10351,6 +10351,32 @@ def test_ReflectionPad_empty(self, device):
             inp = torch.randn(3, 0, 10, 10, device=device)
             mod(inp)
 
+    @onlyOnCPUAndCUDA
+    def test_MultiLabelMarginLoss_empty(self, device):
```
Comment (ngimel) [inline test/test_nn.py:10355]: you are fixing both MultiMarginLoss and MultiLabelMarginLoss, but you are testing only MultiLabelMarginLoss. 
Also, I did not see any changes to cuda, does cuda work already?

**Comment by v0dro** [comment]
@mruberry @ngimel I've added tests for both modules (MultiLabelMarginLoss and MultiMarginLoss) and also updated the CUDA modules (I hadn't added a device option to the module constructor earlier).

Could you have a look again?

File: aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu [hunk 3]
```diff
@@ -58,49 +58,52 @@ void THNN_(MultiLabelMarginCriterion_updateOutput)(
...
-        <<<blocks, threads, 0, c10::cuda::getCurrentCUDAStream()>>>(
-          THCTensor_(data)(state, output),
-          THCTensor_(data)(state, input),
-          THCIndexTensor_(data)(state, target),
-          THCTensor_(data)(state, istarget),
-          nframe, dim,
-          false
-          );
-      THCudaCheck(cudaGetLastError());
+    if (input->numel() != 0) {
```
Comment (mruberry) [inline aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:61]: It seems like this check should be `input->numel() == 0` and go around line 34 to short-circuit quickly and avoid the frees later. That would also minimize the diff size here.

File: aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu [hunk 4]
```diff
@@ -58,49 +58,52 @@ void THNN_(MultiLabelMarginCriterion_updateOutput)(
...
+            nframe, dim,
+            false
+            );
+        THCudaCheck(cudaGetLastError());
+      }
     }
   }
   else {
-    TORCH_INTERNAL_ASSERT(false, "non-empty vector or matrix expected (shouldn't get here)");
+    TORCH_INTERNAL_ASSERT(false, "Expected 2D input with non-zero batch dim, or 1D input with non-zero dim, but got sizes: ",
```
Comment (mruberry) [inline aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:105]: This doesn't seem correct because 2D can have zero batch dim (that's this PR) and 0D tensors also seem OK?

Maybe:

"Expected a tensor with two or fewer dimensions, but got X!"

Also TORCH_CHECK instead of TORCH_INTERNAL_ASSERT.

File: aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu [hunk 5]
```diff
@@ -149,26 +152,29 @@ void THNN_(MultiLabelMarginCriterion_updateGradInput)(
...
-      <<<blocks, threads, 0, c10::cuda::getCurrentCUDAStream()>>>(
-        THCTensor_(data)(state, gradInput),
-        THCTensor_(data)(state, gradOutput),
-        THCTensor_(data)(state, input),
-        THCIndexTensor_(data)(state, target),
-        THCTensor_(data)(state, istarget),
-        gradInput->size(0), gradInput->size(1),
-        reduction == at::Reduction::Mean,
-        reduction != at::Reduction::None);
+    if (input->numel() != 0) {
```
Comment (mruberry) [inline aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:160]: Seems like this check should also live at the start of the function and short-circuit earlier.

File: aten/src/THCUNN/generic/MultiMarginCriterion.cu [hunk 6]
```diff
@@ -59,78 +59,81 @@ void THNN_(MultiMarginCriterion_updateOutput)(
   else if (input->dim() == 2)
   {
     int nframe = input->size(0);
-    THArgCheck(!target->is_empty() && (THTensor_nDimensionLegacyNoScalars(target) == 1) && (THTensor_sizeLegacyNoScalars(target, 0) == nframe), 3,
+    // allow zero-dim target for 2D input.
+    THArgCheck((input->size(1) != 0) && (THTensor_nDimensionLegacyNoScalars(target) == 1) && (THTensor_sizeLegacyNoScalars(target, 0) == nframe), 3,
                "inconsistent target size");
-    dim3 blocks(input->size(0));
-    dim3 threads(MULTIMARGIN_THREADS);
+    if (input->numel() != 0) {
```
Comment (mruberry) [inline aten/src/THCUNN/generic/MultiMarginCriterion.cu:65]: Again: can this check be inverted and a smaller short-circuit added for this case?

File: aten/src/THCUNN/generic/MultiMarginCriterion.cu [hunk 7]
```diff
@@ -59,78 +59,81 @@ void THNN_(MultiMarginCriterion_updateOutput)(
...
-      auto t = THTensor_wrap(output_);
-      auto r = THTensor_wrap(output);
-      at::native::sum_out(r, t, at::IntArrayRef(std::vector<int64_t>{}), false, r.scalar_type());
-      THCTensor_(free)(state, output_);
     }
   }
   else
   {
-    AT_ERROR("non-empty vector or matrix expected, got sizes: ", input->sizes());
+    AT_ERROR("Expected 2D input with non-zero batch dim, or 1D input with non-zero dim, but got sizes: ", input->sizes());
```
Comment (mruberry) [inline aten/src/THCUNN/generic/MultiMarginCriterion.cu:136]: Same updates needed here as mentioned above.

File: aten/src/THCUNN/generic/MultiMarginCriterion.cu [hunk 8]
```diff
@@ -195,44 +198,46 @@ void THNN_(MultiMarginCriterion_updateGradInput)(
   else if (input->dim() == 2)
   {
     int nframe = gradInput->size(0);
-    THArgCheck(!target->is_empty() && (THTensor_nDimensionLegacyNoScalars(target) == 1) && (THTensor_sizeLegacyNoScalars(target, 0) == nframe), 3,
+    THArgCheck((input->size(1) != 0) && (THTensor_nDimensionLegacyNoScalars(target) == 1) && (THTensor_sizeLegacyNoScalars(target, 0) == nframe), 3,
                "inconsistent target size");
-    dim3 blocks(gradInput->size(0));
-    dim3 threads(MULTIMARGIN_THREADS);
+    if (input->numel() != 0) {
```
Comment (mruberry) [inline aten/src/THCUNN/generic/MultiMarginCriterion.cu:203]: Same question here about short-circuiting earlier.

File: aten/src/THCUNN/generic/MultiMarginCriterion.cu [hunk 9]
```diff
@@ -195,44 +198,46 @@ void THNN_(MultiMarginCriterion_updateGradInput)(
...
+        );
+      }
+      THCudaCheck(cudaGetLastError());
     }
-    THCudaCheck(cudaGetLastError());
   }
   else
   {
-    AT_ERROR("non-empty vector or matrix expected, got ", input->sizes());
+    AT_ERROR("Expected 2D input with non-zero batch dim, or 1D input with non-zero dim, but got sizes: ", input->sizes());
```
Comment (mruberry) [inline aten/src/THCUNN/generic/MultiMarginCriterion.cu:240]: Same updates here about error

File: test/test_nn.py [hunk 10]
```diff
@@ -10398,6 +10398,38 @@ def test_ReflectionPad_empty(self, device):
...
+                (torch.nn.MultiLabelMarginLoss().to(device),
+                 torch.randn(0, 10, requires_grad=True, device=device),
+                 torch.ones(0, 10, device=device).type(torch.long))]:
+
+            out = mod(x, y)
+            out.sum().backward()
+
+            self.assertEqual(x, torch.zeros_like(x))
+
+            for p in mod.parameters():
```
[the remark this answers is not in the document]
Comment (v0dro) [reply test/test_nn.py:10417]: Removed the loop.

File: aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu (hunk 3 above)
Comment (v0dro) [reply aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:61]: Thanks for that. Have changed.

File: aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu [hunk 11]
```diff
@@ -149,26 +152,29 @@ void THNN_(MultiLabelMarginCriterion_updateGradInput)(
...
+          THCIndexTensor_(data)(state, target),
+          THCTensor_(data)(state, istarget),
+          gradInput->size(0), gradInput->size(1),
+          reduction == at::Reduction::Mean,
+          reduction != at::Reduction::None);
+    } 
   }
   else {
-    AT_ERROR("non-empty vector or matrix expected, got size: ", gradInput->sizes());
+    AT_ERROR("Expected 2D input with non-zero batch dim, or 1D input with non-zero dim, but got sizes: ", gradInput->sizes());
```
[the remark this answers is not in the document]
Comment (v0dro) [reply aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:177]: Done.

File: aten/src/THCUNN/generic/MultiMarginCriterion.cu (hunk 6 above)
Comment (v0dro) [reply aten/src/THCUNN/generic/MultiMarginCriterion.cu:65]: Done.

File: aten/src/THCUNN/generic/MultiMarginCriterion.cu (hunk 8 above)
Comment (v0dro) [reply aten/src/THCUNN/generic/MultiMarginCriterion.cu:203]: Done.

File: aten/src/THCUNN/generic/MultiMarginCriterion.cu (hunk 7 above)
Comment (v0dro) [reply aten/src/THCUNN/generic/MultiMarginCriterion.cu:136]: Done.

File: test/test_nn.py [hunk 12]
```diff
@@ -10398,6 +10398,38 @@ def test_ReflectionPad_empty(self, device):
...
+                if p.requires_grad:
+                    self.assertEqual(p.grad, torch.zeros_like(p.grad))
+            self.assertEqual(x.grad, torch.zeros_like(x))
+
+            with self.assertRaisesRegex(RuntimeError, 'Expected'):
+                x = torch.randn(0, requires_grad=True)
+                y = torch.ones(10).type(torch.long)
```
[the remark this answers is not in the document]
Comment (v0dro) [reply test/test_nn.py:10424]: Done.

File: aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu (hunk 5 above)
Comment (v0dro) [reply aten/src/THCUNN/generic/MultiLabelMarginCriterion.cu:160]: Done.

File: test/test_nn.py [hunk 13]
```diff
@@ -10398,6 +10398,38 @@ def test_ReflectionPad_empty(self, device):
...
+        for mod, x, y in [
+                (torch.nn.MultiMarginLoss().to(device),
+                 torch.randn(0, 10, requires_grad=True, device=device),
+                 torch.ones(0, device=device).type(torch.long)),
+                (torch.nn.MultiLabelMarginLoss().to(device),
+                 torch.randn(0, 10, requires_grad=True, device=device),
+                 torch.ones(0, 10, device=device).type(torch.long))]:
+
+            out = mod(x, y)
+            out.sum().backward()
```
[the remark this answers is not in the document]
Comment (v0dro) [reply test/test_nn.py:10688]: Not using it throws an error when the tensor is empty.

This is the `out` object:
```
test/test_nn.py tensor([], device='cuda:0', dtype=torch.float32,
       grad_fn=<MultiMarginLossBackward
```

**Comment by v0dro** [comment]
@ngimel could you please have a look at this PR again? CI failures are not related to my changes.

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
Can you please rebase to get a better test signal?

**Comment by ngimel** [comment]
Reverted, it broke asan test.

File: aten/src/ATen/native/LossMultiLabelMargin.cpp [hunk 14]
```diff
@@ -2,6 +2,7 @@
 #include <ATen/AccumulateType.h>
 #include <ATen/Dispatch.h>
 #include <ATen/TensorUtils.h>
+#include "LossMulti.h"
```
Comment (ngimel) [inline aten/src/ATen/native/LossMultiLabelMargin.cpp:5]: Please use standard include here, otherwise builds break (`<ATen/native/LossMulti.h>`)
